### PR TITLE
Update EIP-6551: Preventing circular lock situation with initialize

### PR DESCRIPTION
**Pull request by Abhijaypaliwal**
## Description
This pull request adds the `initialize` function for EIP-6551 as described in the guidelines. The `initialize` function can be called by the owner after ownership change to set the `currOwner` and `prevOwner` variables. Additionally, the `unlockCircularLock` function is introduced to handle situations of circular lock, allowing the previous owner to unlock it.

### Added Functions
- `initialize(uint256 tokenId) external `: Initializes the `currOwner` and `prevOwner` variables after ownership change.

- `unlockCircularLock(uint256 tokenId) external `: Allows the previous owner to unlock the circular lock situation by calling this function.

### Modifications
- Modified the `ERC6551Account` contract to include the new variables `currOwner` and `prevOwner`.

### Tests
- Added comprehensive tests to cover the new functionalities, including the `initialize` and `unlockCircularLock` functions.

## Motivation
This enhancement is in line with the objectives of EIP-6551 and provides additional functionalities for handling ownership transitions and circular lock scenarios.

## Usage
1. After ownership change, the new owner can call the `initialize` function to set the current owner (`currOwner`) and previous owner (`prevOwner`) variables.

2. In cases of circular lock, the previous owner can use the `unlockCircularLock` function to release the lock.

## Issues
The `initialize` function could only be used to change `previousOwner`. and this value could only be used for account recovery. however if new owner of Token transfer it to another owner without initialization, the `previousOwner` is not changed, hence the recovery can be done by third last owner and he can call it without any restriction, which could be malicious.

## Checklist
- [ ] Code follows the EIP-6551 guidelines.
- [ ] Added appropriate tests for the new functionalities.
- [ ] Documentation has been updated to reflect the changes.

### CODE REVIEW

File: EIPS/eip-6551.md [hunk 1]
```diff
@@ -156,6 +156,34 @@ interface IERC6551Account {
      */
     receive() external payable;
 
+    /**
+     * @dev initialize function is used to initialize the tken bound account
```
Comment (auryn-macmillan) [inline EIPS/eip-6551.md:160]: typo: `tken` --> `token`

File: EIPS/eip-6551.md [hunk 2]
```diff
@@ -156,6 +156,34 @@ interface IERC6551Account {
...
+     *  after any ownership change
+     * 
+     *  The function can only be called by current owner of token
+     * 
+     *  The function is used to tract current and previous owner of the NFT
```
Comment (auryn-macmillan) [inline EIPS/eip-6551.md:165]: typo: `tract` --> `track`

File: EIPS/eip-6551.md (hunk 1 above)
Comment (Abhijaypaliwal) [reply EIPS/eip-6551.md:160]: done👍

File: EIPS/eip-6551.md (hunk 2 above)
Comment (Abhijaypaliwal) [reply EIPS/eip-6551.md:165]: done👍

**Comment by jaydenwindle** [comment]
Thanks for the PR @Abhijaypaliwal! This is an interesting concept but seems very opinionated and may not be the best fit for including in the EIP reference implementation. Instead I'd recommend opening up a PR against the [erc6551/reference](https://github.com/erc6551/reference) repo with this functionality as an example in the examples folder. We can work together to polish it and then consider including it in the EIP document.

**Comment by Abhijaypaliwal** [comment]
sure @jaydenwindle i would do the same and we would work together for adding this functionality in near future. I am very much looking forward for this functionality!

**Pull request closed without merging**